plumbing/transport/common.go

```diff
@@ -1,46 +1,114 @@
 // Package clients includes the implementation for different transport protocols
 //
 // go-git needs the packfile and the refs of the repo. The
-// `NewGitUploadPackService` function returns an object that allows to
+// `NewUploadPackService` function returns an object that allows to
 // download them.
 //
 // go-git supports HTTP and SSH (see `Protocols`) for downloading the packfile
 // and the refs, but you can also install your own protocols (see
 // `InstallProtocol` below).
 //
 // Each protocol has its own implementation of
-// `NewGitUploadPackService`, but you should generally not use them
-// directly, use this package's `NewGitUploadPackService` instead.
+// `NewUploadPackService`, but you should generally not use them
+// directly, use this package's `NewUploadPackService` instead.
 package transport
 
 import (
+	"errors"
 	"fmt"
+	"io"
+	"net/url"
+	"regexp"
 
-	"gopkg.in/src-d/go-git.v4/plumbing/transport/common"
-	"gopkg.in/src-d/go-git.v4/plumbing/transport/http"
-	"gopkg.in/src-d/go-git.v4/plumbing/transport/ssh"
+	"gopkg.in/src-d/go-git.v4/plumbing"
 )
 
-// Protocols are the protocols supported by default.
-var Protocols = map[string]common.GitUploadPackServiceFactory{
-	"http":  http.NewGitUploadPackService,
-	"https": http.NewGitUploadPackService,
-	"ssh":   ssh.NewGitUploadPackService,
+var (
+	ErrRepositoryNotFound     = errors.New("repository not found")
+	ErrAuthorizationRequired  = errors.New("authorization required")
+	ErrEmptyUploadPackRequest = errors.New("empty git-upload-pack given")
+	ErrInvalidAuthMethod      = errors.New("invalid auth method")
+)
+
+const (
+	UploadPackServiceName = "git-upload-pack"
+)
+
+// Client can initiate git-fetch-pack and git-send-pack processes.
+type Client interface {
+	// NewFetchPackSession starts a git-fetch-pack session for an endpoint.
+	NewFetchPackSession(Endpoint) (FetchPackSession, error)
+	// NewSendPackSession starts a git-send-pack session for an endpoint.
+	NewSendPackSession(Endpoint) (SendPackSession, error)
+}
+
+type Session interface {
+	SetAuth(auth AuthMethod) error
+	io.Closer
+}
+
+type AuthMethod interface {
+	fmt.Stringer
+	Name() string
+}
+
+// FetchPackSession represents a git-fetch-pack session.
+type FetchPackSession interface {
+	Session
+	// AdvertisedReferences retrieves the advertised references for a
+	// repository. It must be called before Pack.
```

Comment:
> It must be called before Pack.

we should briefly explain why: is this because the protocol says so or because some internal limitation in our code?